tx-pool/src/component/container.rs

```diff
@@ -161,118 +134,90 @@ impl SortedTxMap {
+                .to_owned()
+        }));
+
+        removal.iter().for_each(|e| {
+            self.remove_entry(&e.transaction.proposal_short_id());
+        });
+        removal
     }
 
     pub fn remove_entry(&mut self, id: &ProposalShortId) -> Option<TxEntry> {
-        self.entries.remove(&id).map(|entry| {
-            let deleted = self
-                .sorted_index
-                .remove(&AncestorsScoreSortKey::from(&entry));
-            debug_assert!(deleted, "pending pool inconsistent");
-            // update descendants entries
-            for desc_id in self.get_descendants(&id) {
-                if let Some(key) = self
-                    .entries
-                    .get(&desc_id)
-                    .map(|entry| entry.as_sorted_key())
-                {
-                    self.sorted_index.remove(&key);
-                }
-                if let Some(desc_entry) = self.entries.get_mut(&desc_id) {
-                    // remove entry
-                    desc_entry.sub_entry_weight(&entry);
-                }
-                if let Some(key) = self
-                    .entries
-                    .get(&desc_id)
-                    .map(|entry| entry.as_sorted_key())
-                {
-                    self.sorted_index.insert(key);
-                }
-            }
-            // update links
-            if let Some(link) = self.links.remove(&id) {
-                for p_id in link.parents {
-                    self.links
-                        .get_mut(&p_id)
-                        .map(|link| link.children.remove(&id));
-                }
-                for c_id in link.children {
-                    self.links
-                        .get_mut(&c_id)
-                        .map(|link| link.parents.remove(&id));
-                }
-            }
-            entry
-        })
-    }
+        if !self.contains_key(id) {
+            return None;
+        }
 
-    /// find all ancestors from pool
-    pub fn get_ancestors(&self, tx_short_id: &ProposalShortId) -> HashSet<ProposalShortId> {
-        TxLink::get_ancestors(&self.links, tx_short_id)
-    }
+        let descendants = self.get_descendants(id);
+        let (entry, link) = self.remove(id).expect("checked above");
+
+        // Update its descendants' entries
+        for desc_id in descendants.iter() {
+            let (old_sorted_key, new_sorted_key) = {
+                let desc_entry = self
+                    .get_mut(&desc_id)
+                    .expect("found inconsistency when remove_entry");
+                let old_sorted_key = desc_entry.as_sorted_key();
+
+                desc_entry.sub_entry_weight(&entry);
+                let new_sorted_key = desc_entry.as_sorted_key();
+                (old_sorted_key, new_sorted_key)
+            };
+
+            self.sorted_index.insert(new_sorted_key);
+            self.sorted_index.remove(&old_sorted_key);
```

Comment:
Must remove first, **logically** the `new_sorted_key` may equal to `old_sorted_key`.